### ISSUE TITLE
Default arguments bypass most attributes check (pure, @safe, @nogc)

### ISSUE THREAD

**Issue by dlangBugzillaToGithub** [dlang/dmd#18670]
### CC List

* Andrej Mitrovic (@AndrejMitrovic)
* Walter Bright (@WalterBright)
* Vladimir Panteleev (@CyberShadow)
* Iain Buclaw (@ibuclaw)
* Mathias LANG (@Geod24)

### Description

````markdown
int x = 7;

void foo() pure
{
    // Does not detect use of mutable global and compiles when it shouldn't.
    bar();

    // Correctly detects the use of a mutable global and gives an error
    baz(x);
}

void bar(int a = x) pure {}

void baz(int a) pure {}

void main() {}
````

**Comment by dlangBugzillaToGithub** [comment]
#### andrej.mitrovich (@AndrejMitrovic) commented on 2013-09-16T08:48:37Z



````markdown
So should the declaration of such a function be denied if it's pure, or should only calls be denied where a global is used? E.g.:

// 1. ban this declaration?
void bar(int a = x) pure {}

// 2. or just this call?
bar();

// If #2, then this could be considered ok:
bar(1);
````

**Comment by dlangBugzillaToGithub** [comment]
#### issues.dlang (@jmdavis) commented on 2013-09-16T09:35:22Z



````markdown
The default argument is only a problem if the pure function is called from a pure function. It's perfectly fine if it's called from an impure one. The problem is not that the default argument references a global but that the caller does not detect that the default argument of the function it's calling is a global. So, the caller does not detect that the call violates purity.

The simplest solution is probably to ban the declaration, because then the caller doesn't have to worry about whether the pure function that it's calling has any default arguments which would violate the purity of the caller, but ideally, it's just the call which would be illegal, because the default argument is just fine so long as the caller isn't pure.
````

**Comment by dlangBugzillaToGithub** [comment]
#### pro.mathias.lang (@Geod24) commented on 2020-11-09T00:21:32Z



````markdown
This applies to all attributes but `nothrow`, but only because `nothrow` is the very last to be checked.
E.g. this compiles:
```
int doit() { return 42; }
void f(int i = doit()) nothrow @safe pure @nogc {}
void main() /* nothrow */ @safe pure @nogc { f(); }
```
````

**Comment by dlangBugzillaToGithub** [comment]
#### pro.mathias.lang (@Geod24) commented on 2020-11-09T00:22:24Z



````markdown
*** Issue 13442 has been marked as a duplicate of this issue. ***
````

**Comment by dlangBugzillaToGithub** [comment]
#### ibuclaw (@ibuclaw) commented on 2022-07-16T08:00:04Z



````markdown
*** Issue 19645 has been marked as a duplicate of this issue. ***
````

**Comment by dlangBugzillaToGithub** [comment]
#### ibuclaw (@ibuclaw) commented on 2022-07-16T08:10:58Z



````markdown
I just encountered this when reviewing some @safe code in the dmd compiler implementation.


The crux of which boils down to:
```
int findCondition() @system;

int parseGeneric(int param = findCondition()) @safe
{
    return param;
}

int parseSpec() @safe
{
    return parseGeneric(); // Compiles!
}
```

Call the function directly in @safe code, and you get the expected error.
````

**Comment by dlangBugzillaToGithub** [comment]
#### dlang-bot commented on 2022-07-16T09:16:48Z



````markdown
@ibuclaw created dlang/dmd pull request #14309 "fix Issue 11048 - Default arguments bypass most attributes check (pure, @safe, @nogc)" fixing this issue:

- fix Issue 11048 - Default arguments bypass most attributes check (pure, @safe, @nogc)

https://github.com/dlang/dmd/pull/14309
````

**Comment by TurkeyMan** [comment]
This is 12 years old. I only just noticed this, but now I've done an audit of my code and I found this affects me in several places!
This is a super-critical issue; just trivially and unknowingly violating attributes is not something that can be accepted.

It's hard to imagine what implementation bug could even cause this? Why would a default-arg need special internal logic? Why wouldn't it just lower to a call to the function, and thereby not be any different than if it were called explicitly... what's special about default args that make a lowering inappropriate?

**Comment by dkorpel** [comment]
> Why would a default-arg need special internal logic? Why wouldn't it just lower to a regular call to init the function argument prior to the call?

Because:

> Default parameters are resolved and semantically checked in the context of the function declaration.

https://dlang.org/spec/function.html#function-default-args

It would be really simple to just insert a syntaxCopy() of the default argument at the call site, but that's not according to spec, as it would evaluate to a different value at different usage sites (see the example in the spec).

**Comment by Geod24** [comment]
I think the spec is just poorly worded here, as the example is about visibility. Point 3 seems to hint that this bug is valid, as it deals with a semantic error.

**Comment by dkorpel** [comment]
> Point 3 seems to hint that this bug is valid

Oh there's no question that the bug is valid, I just answered Manu's question why the implementation doesn't / can't treat a default argument exactly the same as writing the expression yourself at the call site.

**Comment by TurkeyMan** [comment]
How could it? The type of the function argument doesn't change at the call-site, unless it's a template, but if the argument is T then it infers the exact return type of the default arg, no?

Are you aware of any particular corner cases?
Maybe the spec is wrong...

**Comment by TurkeyMan** [comment]
Syntax copy at the call-site is the only logical option that anyone would imagine; unless there's some really bad side effect of that, then we're violating the rule of least surprise... and also inviting *super-bad* bugs, as evidenced by the fact this issue exists.

**Comment by TurkeyMan** [comment]
Also you could read that spec sentence several ways; like, the function to be called is called is resolved "in the context of the function declaration" (like, the scope used for resolution and local symbols for template instantiation), but that doesn't mean the actual call must be resolved there. You should be able to get the resolved function SYMBOL, and still put a normal call expression at the call-site... which should confirm attribute correctness?

On a tangent, that line in the spec is the reason `fun(int l = __LINE__, string f = __FILE__)` doesn't work, which is one of the worst common-sense fails that ever happened in D!

**Comment by dkorpel** [comment]
> How could it? The type of the function argument doesn't change at the call-site, unless it's a template, but if the argument is T then it infers the exact return type of the default arg, no?

Because the function declaration and the function call may be in different scopes, identifiers can resolve to different values. 

```D
module m;
enum HELLO = "HELLO";
string getHello(string x = HELLO) => x;
```

```D
import m;

enum HELLO = "GOODBYE"; // hijack default arg
pragma(msg, getHello(HELLO)); // GOODBYE
pragma(msg, getHello()); // HELLO
```

**Comment by dkorpel** [comment]
> On a tangent, that line in the spec is the reason `fun(int l = __LINE__, string f = __FILE__)` doesn't work, which is one of the worst common-sense fails that ever happened in D!

That has always worked. For a long time expressions containing those keywords would fail (e.g. `__FILE__.ptr` would point to the declaration file, not the call site file), but that has been fixed since last year (https://github.com/dlang/dmd/pull/15968). 

Though there's still a 'limitation' that special keywords are resolved eagerly in template arguments, because currently, functions arguments are eagerly type checked at the declaration.

```D
void f(int x = T!__LINE__); // is the type correct? Depends on where it's called from.
```

**Comment by TurkeyMan** [comment]
> > How could it? The type of the function argument doesn't change at the call-site, unless it's a template, but if the argument is T then it infers the exact return type of the default arg, no?
> 
> Because the function declaration and the function call may be in different scopes, identifiers can resolve to different values.

Yeah, but that's what I just said above; it must obviously _resolve_ the symbol in the declaration's scope - it couldn't work any other way, and my reading of the spec is that's what that means.
I don't see how that has anything to do with the call itself though? You take the pre-resolved symbol and insert the call at the callsite. `alias`-es carry around symbols resolved elsewhere all the time; it's extremely normal in D to resolve a symbol in one scope and then transport it around for use somewhere else. I feel like that's all that's happening here?

**Comment by TurkeyMan** [comment]
> For a long time expressions containing those keywords would fail (e.g. `__FILE__.ptr` would point to the declaration file, not the call site file), but that has been fixed since last year ([#15968](https://github.com/dlang/dmd/pull/15968)).

🤯 

I had no idea! That's amazing; and I'm astonished someone had the balls to fix that! That would have surely fallen under the evil category of "mildly-breaking-change" ;)
